src/Compilers/Core/Portable/Serialization/StreamObjectWriter.cs

```diff
@@ -0,0 +1,1459 @@
+                }
+            }
+        }
+
+        private void WriteObjectHeader(object instance, uint memberCount)
+        {
+            _referenceMap.Add(instance);
+
+            _writer.Write((byte)EncodingKind.Object_W);
+
+            Type type = instance.GetType();
+            this.WriteType(type);
+            this.WriteCompressedUInt(memberCount);
+        }
+
+        private static Exception NotWritableException(string typeName)
+        {
+#if COMPILERCORE
+            throw new InvalidOperationException(string.Format(CodeAnalysisResources.NotWritableException, typeName));
+#else
+            throw new InvalidOperationException(string.Format(WorkspacesResources.The_type_0_cannot_be_written_it_does_not_implement_IObjectWritable, typeName));
+#endif
+        }
+
+        // we have s_typeMap and s_reversedTypeMap since there is no bidirectional map in compiler
+        internal static readonly ImmutableDictionary<Type, EncodingKind> s_typeMap = ImmutableDictionary.CreateRange<Type, EncodingKind>(
+            new KeyValuePair<Type, EncodingKind>[]
+            {
+                KeyValuePair.Create(typeof(bool), EncodingKind.BooleanType),
+                KeyValuePair.Create(typeof(char), EncodingKind.Char),
+                KeyValuePair.Create(typeof(string), EncodingKind.StringType),
+                KeyValuePair.Create(typeof(sbyte), EncodingKind.Int8),
+                KeyValuePair.Create(typeof(short), EncodingKind.Int16),
+                KeyValuePair.Create(typeof(int), EncodingKind.Int32),
+                KeyValuePair.Create(typeof(long), EncodingKind.Int64),
+                KeyValuePair.Create(typeof(byte), EncodingKind.UInt8),
+                KeyValuePair.Create(typeof(ushort), EncodingKind.UInt16),
+                KeyValuePair.Create(typeof(uint), EncodingKind.UInt32),
+                KeyValuePair.Create(typeof(ulong), EncodingKind.UInt64),
+                KeyValuePair.Create(typeof(float), EncodingKind.Float4),
+                KeyValuePair.Create(typeof(double), EncodingKind.Float8),
+                KeyValuePair.Create(typeof(decimal), EncodingKind.Decimal),
+            });
+
+        internal static readonly ImmutableDictionary<EncodingKind, Type> s_reverseTypeMap = s_typeMap.ToImmutableDictionary(kv => kv.Value, kv => kv.Key);
+
+        // byte marker for encoding compressed uint
+        internal static readonly byte ByteMarkerMask = 3 << 6;
+        internal static readonly byte Byte1Marker = 0;
+        internal static readonly byte Byte2Marker = 1 << 6;
+        internal static readonly byte Byte4Marker = 2 << 6;
+
+        /// <summary>
+        /// The encoding prefix byte used when encoding <see cref="Variant"/> values.
+        /// </summary>
+        internal enum EncodingKind : byte
+        {
+            Null,
+            Type,
+            TypeRef,      // type ref id as 4 bytes 
+            TypeRef_B,    // type ref id as 1 byte
```

Comment:
1. Instead of comments, the names could be clearer.  i.e. TypeRefId_SingleByte.
2. These comments don't show up in IDE features.  Please use doc comment for documenting symbols.